testing/unittests/metrics/MetricFrameworkTests.cpp

```diff
@@ -0,0 +1,190 @@
+
+        //
+        // Test increment by > 1
+        pCounter->inc(2);
+        CPPUNIT_ASSERT_EQUAL_MESSAGE("Counter metric expected value to be 3 after increment by 2", 3, static_cast<int>(pCounter->queryValue()));
+    }
+
+
+    void Test_gauge_metric_updates_properly()
+    {
+        std::shared_ptr<GaugeMetric> pGauge = std::make_shared<GaugeMetric>("test-gauge", "description");
+        CPPUNIT_ASSERT_EQUAL_MESSAGE("Gauge metric expected initial value to be 0", 0, static_cast<int>(pGauge->queryValue()));
+
+        //
+        // Test initial setting of gauge
+        pGauge->set(25);
+        CPPUNIT_ASSERT_EQUAL_MESSAGE("Gauge metric expected value to be 25 after setting", 25, static_cast<int>(pGauge->queryValue()));
+
+        //
+        // Test updating gauge
+        pGauge->add(10);
+        CPPUNIT_ASSERT_EQUAL_MESSAGE("Gauge metric expected value to be 35", 35, static_cast<int>(pGauge->queryValue()));
+
+        pGauge->add(-5);
+        CPPUNIT_ASSERT_EQUAL_MESSAGE("Gauge metric expected value to be 30", 30, static_cast<int>(pGauge->queryValue()));
+    }
+
+
+    void Test_custom_metric()
+    {
+        int customCounter = 0;
+        std::shared_ptr<CustomMetric<int>> pCustomCounter = std::make_shared<CustomMetric<int>>("custom-counter", "description", METRICS_COUNTER, customCounter);
+        CPPUNIT_ASSERT_EQUAL_MESSAGE("Counter metric expected initial value to be 0", 0, static_cast<int>(pCustomCounter->queryValue()));
+
+        customCounter++;
+        CPPUNIT_ASSERT_EQUAL_MESSAGE("Counter metric expected value to be 1 after default increment", 1, static_cast<int>(pCustomCounter->queryValue()));
+    }
+
+
+    void Test_reporter_calls_sink_to_start_and_stop_collection()
+    {
+        frameworkTestReporter.startCollecting();
+        CPPUNIT_ASSERT_EQUAL_MESSAGE("Test sink start collection was not called", true, pTestSink->isCollecting);
+        frameworkTestReporter.stopCollecting();
+        CPPUNIT_ASSERT_EQUAL_MESSAGE("Test sink stop collection was not called", false, pTestSink->isCollecting);
+    }
+
+
+    void Test_reporter_manages_metrics_properly()
+    {
+        int numAdded;
+        std::shared_ptr<CounterMetric> pCounter = std::make_shared<CounterMetric>("test-counter", "description");
+        std::shared_ptr<GaugeMetric> pGauge = std::make_shared<GaugeMetric>("test-gauge", "description");
+        frameworkTestReporter.addMetric(pCounter);
+        frameworkTestReporter.addMetric(pGauge);
+        numAdded = 2;
+
+        frameworkTestReporter.startCollecting();
+
+        //
+        // Make sure the initial list is correct
+        auto numMetrics = frameworkTestReporter.queryMetricsForReport("testsink").size();
+        CPPUNIT_ASSERT_EQUAL_MESSAGE(VStringBuffer("Expected %d metrics to be returned", numAdded).str(), numAdded, static_cast<int>(numMetrics));
```

Comment:
This is a good example of an assert that is much harder to read because of the message.

CPPUNIT_ASSERT_EQUAL(numAdded, static_cast<int>(numMetrics));

even better if numAdded had the same type (size_t?) so there would be no need for the cast.

Reply:
I removed all message asserts along with all explicit casts of test values.